Datasources import of JSON schema

carbonarok opened the pull request:
<!--
    Thank you for your interest in contributing to Nautobot! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->
### Fixes: #<ISSUE NUMBER GOES HERE>
<!--
    Please include a summary of the proposed changes below.
-->

jathanism committed:
Assert that tags persist in changelog on API PATCH requests (#558)

* Assert that tags persist in changelog on API PATCH requests
* This also adds a regression test for API PATCH requests (partial updates) for changelogs.

Co-authored-by: Glenn Matthews <glenn.matthews@networktocode.com>

jathanism committed:
Add changelog for #499

Joezeppe committed:
Closes #530: Improve serializer docstring. (#566)

Co-authored-by: Sammy Chafiqui <sammy.chafiqui@gmail.com>

jathanism committed:
Add changelog for #530

glennmatthews committed:
Support device_types metadata key when importing ConfigContexts from Git. Fixes #562 (#565)

glennmatthews committed:
Add release-note for #562

smk4664 committed:
Give the user some help on empty repos. (#535)

Move the commit.hexsha check as an empty repo won't have a commit.
This allows the user to fix the repo without needing to delete the git
repository record.

In python, the second check in an and clause is not evaluated
if the first check fails.

jathanism committed:
Add changelog for #421

smk4664 committed:
Add autodetection of mime_type on git datasourced export_template. (#563)

This change tries to auto-detect the mime_type from the export template extension. If not found, defaults to `text/plain`.

jathanism committed:
Add changelog for #561

glennmatthews committed:
Port of netbox-community/netbox#5704 - allow 'None' to be selected on various form fields (#583)

glennmatthews committed:
Add release-notes for #583

whitej6 committed:
resolves #342 issue with emitting null time_zone attr on Site object. (#419)

* Fix #342 Issue with emitting null time_zone attr on Site object.

- Refactor `Site.time_zone` just use what comes with `django-timezone-field`
- Removed `nautobot.core.api.fields.TimeZoneField`
- Replaced custom `TimeZoneField` with stock `timezone_field.rest_framework.TimeZoneSerializerField`
  - Enabled `allow_null=True` so that `null` can be provided as a value to this field
- Refactored new unit tests to have `time_zone=None` be happy and `time_zone=""` to be sad as this is in line w/ the default behavior from `django-timezone-field`, asserts consistency, and eliminates the need for us to maintain more custom code

Co-authored-by: Glenn Matthews <glenn.matthews@networktocode.com>
Co-authored-by: jathanism <jathan@gmail.com>

jathanism committed:
Add changelog for #342

carbonarok committed:
Inital commit

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+
+def refresh_git_config_context_schemas(repository_record, job_result, delete=False):
+    """Callback function for GitRepository updates - refresh all ConfigContext records managed by this repository."""
+    if "extras.configcontextschema" in repository_record.provided_contents and not delete:
```
Thoughts on whether it's desirable to have config-context-schemas as a separate per-repository toggle from config-contexts alone? Or should we keep it simple and have both config-contexts and config-context-schemas enabled or disabled for a repo by a single shared toggle?

glennmatthews replied:
Let's keep them separately addressable for now - thanks.

carbonarok committed:
Black .

glennmatthews committed:
Add custom links and plugin template content to JobResult detail views (#576)

glennmatthews committed:
Add release-notes for #576

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+
+
+def refresh_git_config_context_schemas(repository_record, job_result, delete=False):
+    """Callback function for GitRepository updates - refresh all ConfigContext records managed by this repository."""
```
```suggestion
    """Callback function for GitRepository updates - refresh all ConfigContextSchema records managed by this repository."""
```

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+
+    managed_config_context_schemas = set()
+
+    # First, handle the "flat file" case - data files in the root config_context_path,
+    # whose metadata is expressed purely within the contents of the file:
```
I don't see code to support the "second" option that ConfigContext's currently support, i.e. `config_contexts/<filter_type>/<slug>.(json|yaml)`. Is that an intentional omission?

carbonarok replied:
`yaml.safe_load` loads JSON files.

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+
+def update_git_config_context_schemas(repository_record, job_result):
+    """Refresh any config context schemas provided by this Git repository."""
+    config_context_path = os.path.join(repository_record.filesystem_path, "config_context_schemas")
```
Let's discuss design a bit - does it make sense for schemas to be stored in a separate directory path from the contexts themselves, or should they be in the same directory hierarchy? I'm not sure, but thought it's worth considering.

glennmatthews replied:
Separate directories is fine for now, I think.

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+                try:
+                    context_schema_data = yaml.safe_load(fd)
+                except Exception as exc:
+                    raise RuntimeError(f"Error in loading config context data from `{file_name}`: {exc}")
```
```suggestion
                    raise RuntimeError(f"Error in loading config context schema data from `{file_name}`: {exc}")
```

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+                managed_config_context_schemas.add(context_name)
+            else:
+                raise RuntimeError(
+                    f"Error in loading config context data from `{file_name}`: data must be a dict or list of dicts"
```
```suggestion
                    f"Error in loading config context schema data from `{file_name}`: data must be a dict or list of dicts"
```

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+            if isinstance(context_schema_data, dict):
+                context_name = import_config_context_schema(context_schema_data, repository_record, job_result, logger)
+                managed_config_context_schemas.add(context_name)
+            else:
```
Not supporting the option of defining multiple schemas in a single file as a list of dicts?

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+            job_result.log(
+                str(exc),
+                level_choice=LogLevelChoices.LOG_FAILURE,
+                grouping="config contexts",
```
```suggestion
                grouping="config context schemas",
```

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+    created = False
+    modified = False
+
+    if not context_schema_data.get("title"):
```
Why mismatching the field names (`title` vs. `name`, `properties` vs `data_schema`)?

carbonarok replied:
I was following the example file @lampwins  sent on slack
``` json
{
    "$id": "https://example.com/person.schema.json",
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "title": "Person",
    "type": "object",
    "properties": {
        "firstName": {
            "type": "string",
            "description": "The person's first name."
        },
        "lastName": {
            "type": "string",
            "description": "The person's last name."
        },
        "age": {
            "description": "Age in years which must be equal to or greater than zero.",
            "type": "integer",
            "minimum": 0
        }
    }
}
```

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+    if created:
+        schema_record.validated_save()
+        job_result.log(
+            "Successfully created config context",
```
```suggestion
            "Successfully created config context schema",
```

Same suggestion for the other log messages below.

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+            "Successfully created config context",
+            obj=schema_record,
+            level_choice=LogLevelChoices.LOG_SUCCESS,
+            grouping="config contexts",
```
```suggestion
            grouping="config context schemas",
```

Same suggestion below.

glennmatthews commented on nautobot/extras/tests/test_datasources.py:
```diff
@@ -71,6 +72,27 @@ def setUp(self):
             job_id=uuid.uuid4(),
         )
 
+        self.config_context_schema = {
+            "title": "Person Schema",
+            "description": "Schema for defining first names, last names and ages.",
+            "type": "object",
+            "properties": {
+                "firstName": {
+                    "type": "string",
+                    "description": "The person's first name.",
+                },
+                "lastName": {
+                    "type": "string",
+                    "description": "The person's last name.",
+                },
+                "age": {
+                    "description": "Age in years which must be equal to or greater than zero.",
+                    "type": "integer",
+                    "minimum": 0,
+                },
+            },
+        }
```
This data structure doesn't seem to align well with how config contexts themselves are defined in Git. I would have expected something more like the following (with `_metadata` tracking all of the fields except the schema itself, and the data-schema as a first-class part of the data model:

```suggestion
        self.config_context_schema = {
            "_metadata": {
                "title": "Person Schema",
                "description": "Schema for defining first names, last names and ages.",
                "type": "object",
            },
            "firstName": {
                "type": "string",
                "description": "The person's first name.",
            },
            "lastName": {
                "type": "string",
                "description": "The person's last name.",
            },
            "age": {
                "description": "Age in years which must be equal to or greater than zero.",
                "type": "integer",
                "minimum": 0,
            },
        }
```

Definitely open to discussion on this point though.

lampwins replied:
One thing to consider here is the `data_schema` might not be the only bit of non-metadata information we store for the ConfigContextSchema in the future. There are several frontend libraries out there that support rendering dynamic forms from json schemas. A couple (including the [most popular](https://jsonforms.io/)) make use of an additional UI schema which dictates _how_ certain parts of the json schema should be rendered. Anticipating this potential future use case is why I named the `data_schema` field as I did.

glennmatthews replied:
Fair points! Would definitely appreciate your input on what would make the most sense for the file structure then.

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,152 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+        if schema_record.name not in preserve:
+            schema_record.delete()
+            job_result.log(
+                f"Deleted config context {schema_record}",
+                level_choice=LogLevelChoices.LOG_WARNING,
+                grouping="config contexts",
```
```suggestion
                f"Deleted config context schema {schema_record}",
                level_choice=LogLevelChoices.LOG_WARNING,
                grouping="config context schemas",
```

glennmatthews reviewed (commented):
General comments:

- Needs documentation added to `nautobot/docs/user-guides/git-data-source.md` and `nautobot/docs/models/extras/gitrepository.md`
- I don't see a mechanism for associating config-contexts and config-context-schemas together via Git?
  - For added bonus points, any association between the two should be validated at sync time, i.e. if a config-context is updated such that it no longer complies with the provided schema, the update should fail with an appropriate error log.

jvanderaa committed:
Adds Copy Button (#144)

* Adds basic copy button

* Updates to use MDI buttons

* Created btn-inline class for text buttons inline to not bloat text.

* Adds basic copy button

* Updates to use MDI buttons

* Created btn-inline class for text buttons inline to not bloat text.

* Small iteration update to include inline and hover

* Fixes broken from merge conflict; hover on title.

* Moves into base css, removes extra span

* Update nautobot/dcim/templates/dcim/device/base.html

Co-authored-by: Glenn Matthews <glenn.matthews@networktocode.com>

* Update nautobot/dcim/templates/dcim/device.html

Co-authored-by: Glenn Matthews <glenn.matthews@networktocode.com>

* Update nautobot/dcim/templates/dcim/device.html

Co-authored-by: Glenn Matthews <glenn.matthews@networktocode.com>

* Update nautobot/dcim/templates/dcim/device.html

Co-authored-by: Glenn Matthews <glenn.matthews@networktocode.com>

Co-authored-by: Bryan Culver <me@bryanculver.com>
Co-authored-by: Glenn Matthews <glenn.matthews@networktocode.com>

glennmatthews committed:
Add release-notes for #143

u1735067 committed:
Fix git imported template immediately removed during import (#582)

* Fix git template import for some objects

Closes https://github.com/nautobot/nautobot/issues/570.

* Add test for Git imported vlan template

glennmatthews committed:
Add release-notes for #570

scetron committed:
Update VLAN Count in VLAN Groups page (#581)

* update vlan count var for accurate count

* Update nautobot/ipam/views.py

Co-authored-by: Glenn Matthews <glenn.matthews@networktocode.com>

glennmatthews committed:
Add release-note for #490

nniehoff committed:
Example django-health-check (#302)

This PR adds `django-health-check` and publishes a `/health/` endpoint. It also introduces and documents the `nautobot-server health_check` command.

Co-authored-by: Jathan McCollum <jathan@gmail.com>

jathanism committed:
Add changelog for #297 (health check)

carbonarok committed:
Minor string fixes.

jathanism committed:
Prepare Release v1.0.3 (#603)

* Prepare Release v1.0.3

- Bumped version to v1.0.3
- Updated release date in changelog

jathanism committed:
Bump version and release notes for v1.0.4b1

jathanism committed:
Fix #494 - Objects with `status` fields now emit limited choices (#604)

- This adds a `get_queryset()` method to `nautobot.extras.api.fields.StatusSerializerField`
- New specialized `test_status_options_returns_expected_choices()` method added to `nautobot.utilities.testing.api.APIVIewTestCases` to run on all API test cases
  - This test will be skipped of an test case does not have a `status` field returned in its metadata so a larger amount of skipped tests is expected (69 as of this writing)

jathanism committed:
Add changelog for #494 (`status` choices in `OPTIONS` requests)

carbonarok committed:
Enable new navbar for plugins (#572)

* Inital commit

* Configured navigation for all nautobot apps.

* Black formating

* Added weight to items. Created button subclass.

* Link reverse sanity check.

* Black and added button class default.

* Added docstrings. Moved nautobot_app to app. Added PermissionsMixin. Added isinstance and issubclass checking. Added new choiceset for button colors.

* Added docstring to NautobotConfig

* App class name change to apps. Improved register logic. Reverted from item list to item dict.

* Black

* Improved if statements.

* Added button weight processing. Moved plugin loading into authenicated if statement. Added new templatetag to return True if user has at least one permission.

* Added documentation to registry. Reverted poetry lock. Added button icon choiceset.

* Removed duplicates from permissions.

* Added logging when objects have already been defined.

* Black

* Added testing. Added function just checking if navbar object already exists.

* Added test case. Modified version doc. Permissions changed to set. Added error if failed to login with test suite.

* Black

* Revert docker file changes. Added permissions checking on test_navbar_render_restricted_ui. Fixed navigation wrong link.

* Changed base class to use abc. fixed some navigation wrong links and permissions. added more choices to button choicesets.

* Fixed cable permissions. Reverted changes to poetry lock. Fixed spelling in doc string.

* Updated to registry documentation.

* Update registry doc.

* Added nautobot app base class to plugin class. Added register nav menu function to plugin initialisation. Changed plugin nav menu register function name. Added test to dummy plugin. Added modified tab to dummy plugin.

* Black

* Minor documentation changes. Removed unused testcase. Minor modifications to dummy plugin. Plugin registers navbar through using plugin objects to convert into NavMenu objects.

* Black

* Moved documentation to new file in /docs/development/

* Changes to docs. Added type checking on naviagtion objects.

* Added plugin testcases for navbar generation.

* Minor documentation fix.

* Black

* Removed redundant file. Removed redundant code. Fixed doc strings.

* Added weights back to fixed fields

* F strings removed.

* Added tab for tip.

* NavMenuItem.link_text chnages to name

glennmatthews committed:
Ensure that deleting a record deletes its associated RelationshipAssociations (#593)

* Fix #460: ensure that deleting a record deletes its associated RelationshipAssociations

* Add related_query_name to GenericRelations and associated testing.

* Add keepdb option when running tests as a whole

glennmatthews committed:
Add release-note for #460

lampwins committed:
Implements #274 - config context schemas (#579)

* closes #274 - config context schemas

* resolve poetry.lock conflict

* added unit tests

* added integration tests using the splinter library and created the SplinterTestCase base class

* added docs

* fix integration tests

* pr feedback

* add object based schema validation checks

* pr feedback

* fix view tests

* added validation tab

* added api and view tests for device/vm/config context

* pr updates

lampwins committed:
add #274 release notes

lampwins committed:
fix config context schema nav menu link name

jathanism committed:
Merge remote-tracking branch 'upstream' into develop-1.1.0

jathanism committed:
Fix the bad merge resolution I did in develop merge.

carbonarok committed:
Added weight feature to git repos. Added documentation. Added mutliple schemas per file processing. Modified data collection from schema files.

carbonarok committed:
Removed redundant comments.

glennmatthews commented on nautobot/docs/models/extras/gitrepository.md:
````diff
@@ -114,6 +114,79 @@ Files in a `config_contexts/devices/` and/or `config_contexts/virtual_machines/`
+  context_schema_2.yaml   # YAML data will be imported as-is, with scoping derived from its contents
+```
+
+When loading the schema, the key `_metadata` will be extracted from the loaded data and used to define the config context schemer's metadata; all remaining data will form the config context data schema.
````
While this follows the existing pattern established by config-contexts, I think we should consider https://github.com/lampwins/nautobot/pull/1#discussion_r656489621 - do we want to plan ahead for the possibility of the config context schema eventually including additional data fields beyond just the data schema? In other words, should we consider instead of:

```json
{
  "_metadata": {
    "name": "Config Context Schema 1",
    "description": "Schema for defining first names, last names and ages.",
    "config_contexts": [
      {"name": "Config Context 1"},
    ],
  },
  // data schema follows
```

possibly something like:

```json
{
  "name": "Config Context Schema 1",
  "description": "Schema for defining first names, last names and ages.",
  "config_contexts": [
     {"name": "Config Context 1"},
  ],
  "data_schema": {...}
  // future attributes to be added here, perhaps?
```

glennmatthews commented on nautobot/docs/models/extras/gitrepository.md:
```diff
@@ -114,6 +114,79 @@ Files in a `config_contexts/devices/` and/or `config_contexts/virtual_machines/`
+
+When loading the schema, the key `_metadata` will be extracted from the loaded data and used to define the config context schemer's metadata; all remaining data will form the config context data schema.
+
+Inside of the metadata there is a list called `config_contexts`, this holds all of the config contexts the schema applies too. The config contexts are defined through dictionaries with attributes which can identify the context. Below shows that the schema applies to a config context with a name of `"Config Context 1"`.
```
Interesting design, particularly in that it inverts the internal data model (in which config contexts reference the schema they use via a ForeignKey, rather than the schema directly listing the contexts it applies to). Does modeling it this way make it easier to implement and maintain versus having `schema` as a `_metadata` field on the `config_context` file(s)?

One small potential downside I see here is that if a user wants to add a new config context that uses an existing schema, the changeset isn't localized to just the config context file - they have to modify the schema file as well to reference the new config context. The alternate approach would keep the changes isolated, maybe worth considering?

carbonarok replied:
I agree, i dont think this makes much sense to change the config schema everytime. 
How do you suggest we fix this? My thoughts would be to change the processing so that the config context has the schema set in its metadata 
Sound good to you @glennmatthews ?

glennmatthews commented on nautobot/docs/user-guides/git-data-source.md:
```diff
@@ -243,6 +255,18 @@ Here's an example, with some of the details omitted for brevity.
 
+### Config Context Schemas
+
+Detailed information on [config context schema](../models/extras/gitrepository.md#configuration-context-schemas) in Git Repositories.
```
```suggestion
Detailed information on [config context schemas](../models/extras/gitrepository.md#configuration-context-schemas) in Git Repositories.
```

glennmatthews commented on nautobot/docs/user-guides/git-data-source.md:
```diff
@@ -243,6 +255,18 @@ Here's an example, with some of the details omitted for brevity.
+
+Detailed information on [config context schema](../models/extras/gitrepository.md#configuration-context-schemas) in Git Repositories.
+
+Config context schemas are used to enforce data validation on config contexts. These schema are managed via the config [context schema model](../models/extras/configcontextschema.md) and are optionally linked to config context instances, in addition to devices and virtual machines for the purpose of validating their local context data.
```
```suggestion
Config context schemas are used to enforce data validation on config contexts. These schema are managed via the [config context schema model](../models/extras/configcontextschema.md) and are optionally linked to config context instances, in addition to devices and virtual machines for the purpose of validating their local context data.
```

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,194 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+            continue
+        job_result.log(
+            f"Loading config context schema from `{file_name}`",
+            grouping="config contexts",
```
```suggestion
            grouping="config context schemas",
```

glennmatthews commented on nautobot/extras/datasources/git.py:
```diff
@@ -551,6 +553,194 @@ def delete_git_config_contexts(repository_record, job_result, preserve=(), prese
+            if isinstance(context_schema_data, dict):
+                context_name = import_config_context_schema(context_schema_data, repository_record, job_result, logger)
+                managed_config_context_schemas.add(context_name)
+            elif isinstance(context_schema_data, list):
```
Need to document this option

glennmatthews commented on nautobot/extras/tests/test_datasources.py:
```diff
@@ -1,4 +1,5 @@
 import json
+from nautobot.extras.models.models import ConfigContextSchema
```
This should be moved down to be grouped with the rest of `from nautobot.extras.models import...`

glennmatthews commented on nautobot/extras/tests/test_datasources.py:
```diff
@@ -204,6 +235,18 @@ def populate_repo(path, url):
+                config_context_schema_metadata = config_context_schema.pop("_metadata", {})
+                self.assertIsNotNone(config_context_schema_record)
+                self.assertEqual(config_context_schema_metadata["name"], config_context_schema_record.name)
+                self.assertEqual(config_context_schema, config_context_schema_record.data_schema)
```
Should also test to verify that the schema-to-config-context mapping is also set up properly.

carbonarok committed:
Added new key 'data_schema' to data structure.

carbonarok committed:
Moved schema processing to config context

carbonarok committed:
Merge branch '274-config-context-schemas' of https://github.com/lampwins/nautobot into kc_datasources_json_schema

carbonarok committed:
Merge branch 'develop-1.1.0' into kc_datasources_json_schema

carbonarok committed:
Black and document fix.

carbonarok committed:
Updated testing for datasource class. Removed setdefaults.

carbonarok committed:
Variable name change
and added secondary order using name in lamdba

carbonarok committed:
Black

carbonarok committed:
Modified documentation. Added dict isinstance check. Removed unused import.

carbonarok committed:
Modified docs. Added slots to registry class. Modified testing.

carbonarok committed:
Changes to documentation according to comments.

carbonarok committed:
Removed id and schema from examples.

carbonarok committed:
Fix KeyError schema.
